### DISCUSSION TITLE
Function::Parameters in Mojo

### DISCUSSION THREAD

**Discussion by david-dot-krupicka** [mojolicious/mojo#2130] [Ideas]
Hello,

Although I don't like keyword `fun` and `method` as an addition to `sub`, the module has certain benefits.
If you forget to add some parameter, it fails at compile time. And shifting to `$self` is automatic.

Could it be used within Mojo, or is it absolute nonsense? If yes, `has` would need to be somehow adapted.

Best regards
David

**Comment by jberger** [comment]
Although your application classes must inherit from certain classes in most cases (ie, application, controllers, plugins, etc.) Mojolicious doesn't specify what tooling you use to build your classes or create methods. As long as the inheritance tree works and methods are implemented as documented, any tooling should work. Moo(se), Function::Parameters, Moops, etc.

(Note: Prototype-based ones might take a little tinkering, but no one asked that here :) )

**Reply by david-dot-krupicka** [reply]
Hi Joel, thank you for the answer.
Could you provide a small example, please? And yes, I would like to ask about the tinkering as well. :-)

**Reply by jberger** [reply]
Personally, the (new) native parameters are usually good enough for me. Unfortunately I don't have time to put together an example using a library I don't use. Please, try it for yourself and please post if you have troubles.
